### PR TITLE
Refactoring code before Nightshade PR

### PR DESCRIPTION
**Pull request by ilblackdragon**
- Move protos, primitves, storage to core folder.
 - Storage renamed to store, Store has StoreUpdate that accumulates updates across all of the changes and commits them in one step, which avoids invalid state in the DB in case of crashes or exits.
 - Added near/config for new Genesis config definition. Removed BLS signatures from the configuration.
 - Base 58 switched everywhere to Base 64 - we used both but it creates confusion.

Step 1 of #928 merge

### CODE REVIEW

File: core/primitives/src/crypto/signer.rs [hunk 1]
```diff
@@ -84,28 +86,32 @@ pub fn get_key_file(key_store_path: &Path, public_key: Option<String>) -> KeyFil
...
+    #[serde(with = "bs64_serializer")]
     pub public_key: PublicKey,
-    #[serde(with = "bs58_serializer")]
+    #[serde(with = "bs64_serializer")]
     pub secret_key: SecretKey,
-    #[serde(with = "bs58_serializer")]
-    pub bls_public_key: BlsPublicKey,
-    #[serde(with = "bs58_serializer")]
-    pub bls_secret_key: BlsSecretKey,
+    //    #[serde(with = "bs64_serializer")]
```
Comment (bowenwang1996) [inline core/primitives/src/crypto/signer.rs]: ?

File: core/primitives/src/crypto/signer.rs [hunk 2]
```diff
@@ -84,28 +86,32 @@ pub fn get_key_file(key_store_path: &Path, public_key: Option<String>) -> KeyFil
...
+    //    pub bls_secret_key: BlsSecretKey,
 }
 
 pub fn write_block_producer_key_file(
     key_store_path: &Path,
     public_key: PublicKey,
     secret_key: SecretKey,
-    bls_public_key: BlsPublicKey,
-    bls_secret_key: BlsSecretKey,
+    //    bls_public_key: BlsPublicKey,
```
Comment (bowenwang1996) [inline core/primitives/src/crypto/signer.rs]: ?

File: core/primitives/src/crypto/signer.rs [hunk 3]
```diff
@@ -162,43 +168,50 @@ pub fn get_or_create_key_file(
 ) -> BlockProducerKeyFile {
     if !key_store_path.exists() {
         let (public_key, secret_key) = get_key_pair();
-        let bls_secret_key = BlsSecretKey::generate();
-        let bls_public_key = bls_secret_key.get_public_key();
+        //        let bls_secret_key = BlsSecretKey::generate();
```
Comment (bowenwang1996) [inline core/primitives/src/crypto/signer.rs]: ?

File: core/primitives/src/crypto/signer.rs [hunk 4]
```diff
@@ -233,13 +250,13 @@ impl EDSigner for InMemorySigner {
...
-    #[inline]
-    fn bls_public_key(&self) -> BlsPublicKey {
-        self.bls_public_key.clone()
-    }
-
-    fn bls_sign(&self, data: &[u8]) -> PartialSignature {
-        self.bls_secret_key.sign(data)
-    }
-}
+//impl BLSSigner for InMemorySigner {
```
Comment (bowenwang1996) [inline core/primitives/src/crypto/signer.rs]: ?

File: core/primitives/src/network.rs [hunk 5]
```diff
@@ -231,10 +212,11 @@ impl TryFrom<network_proto::PeerMessage> for PeerMessage {
                     .collect::<Result<Vec<_>, _>>()?;
                 Ok(PeerMessage::InfoGossip(peer_info))
             }
-            Some(network_proto::PeerMessage_oneof_message_type::message(message)) => {
-                Ok(PeerMessage::Message(message))
-            }
+//            Some(network_proto::PeerMessage_oneof_message_type::message(message)) => {
```
Comment (bowenwang1996) [inline core/primitives/src/network.rs]: ?

File: core/primitives/src/network.rs [hunk 6]
```diff
@@ -246,17 +228,18 @@ impl From<PeerMessage> for network_proto::PeerMessage {
...
                         peers_info.into_iter().map(std::convert::Into::into),
                     ),
                     ..Default::default()
                 };
                 Some(network_proto::PeerMessage_oneof_message_type::info_gossip(gossip))
             }
-            PeerMessage::Message(message) => {
-                Some(network_proto::PeerMessage_oneof_message_type::message(message))
-            }
+//            PeerMessage::Message(message) => {
```
Comment (bowenwang1996) [inline core/primitives/src/network.rs]: ?

File: core/primitives/src/test_utils.rs [hunk 7]
```diff
@@ -41,13 +55,13 @@ pub fn get_bls_key_pair_from_seed(seed_string: &str) -> (BlsPublicKey, BlsSecret
 impl InMemorySigner {
     pub fn from_seed(account_id: &str, seed_string: &str) -> Self {
         let (public_key, secret_key) = get_key_pair_from_seed(seed_string);
-        let (bls_public_key, bls_secret_key) = get_bls_key_pair_from_seed(seed_string);
+//        let (bls_public_key, bls_secret_key) = get_bls_key_pair_from_seed(seed_string);
```
Comment (bowenwang1996) [inline core/primitives/src/test_utils.rs]: ?

File: runtime/wasm/src/prepare.rs [hunk 8]
```diff
@@ -138,31 +118,29 @@ impl<'a> ContractModule<'a> {
...
-
-			let ext_func = env
-				.funcs
-				.get(import.field().as_bytes())
-				.ok_or_else(|| Error::Instantiate)?;
-			if !ext_func.func_type_matches(func_ty) {
-				return Err(Error::Instantiate);
-			}
-			*/
+            /*
```
Comment (bowenwang1996) [inline runtime/wasm/src/prepare.rs]: ?

**Comment by vans163** [comment]
Should this be URL and filename safe alphabet + padding false base64?

File: runtime/wasm/src/prepare.rs (hunk 8 above)
Comment (ilblackdragon) [reply runtime/wasm/src/prepare.rs]: Just reformatted the file, this was here before.

File: core/primitives/src/crypto/aggregate_signature.rs [hunk 9]
```diff
@@ -276,19 +275,19 @@ impl<E: Engine> ToBytes for Signature<E> {
 }
 
 impl<E: Engine> TryFrom<&[u8]> for Signature<E> {
-    type Error = Box<Error>;
```
Comment (nearmax) [inline core/primitives/src/crypto/aggregate_signature.rs]: This PR replaces many `type Error = Box<std::error::Error>;` with `type Error = String;`. What is the reason? We should using `type Error = Box<std::error::Error>;` wherever it is possible, and moving our code in that direction, not the other way around.

File: core/primitives/Cargo.toml [hunk 10]
```diff
@@ -1,13 +1,12 @@
 [package]
-name = "primitives"
+name = "near-primitives"
```
Comment (nearmax) [inline core/primitives/Cargo.toml]: What is the reason for that?

File: core/primitives/src/crypto/signature.rs [hunk 11]
```diff
@@ -220,7 +236,19 @@ impl fmt::Display for SecretKey {
...
 impl<'a> From<&'a Signature> for String {
     fn from(h: &'a Signature) -> Self {
-        bs58::encode(h).into_string()
+        base64::encode(h)
+    }
+}
+
+impl<'a> From<&'a Signature> for Vec<u8> {
+    fn from(h: &'a Signature) -> Self {
+        (h.0).0.to_vec()
```
Comment (nearmax) [inline core/primitives/src/crypto/signature.rs]:
```suggestion
        h.0.0.to_vec()
```
And below too

File: core/primitives/src/sharding.rs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        return fetched_parts;
+    }
+
+    pub fn reconstruct(&mut self, data_shards: usize, parity_shards: usize) {
+        let rs = ReedSolomon::new(data_shards, parity_shards).unwrap();
+        rs.reconstruct_shards(self.parts.as_mut_slice()).unwrap();
+    }
+
+    pub fn get_merkle_hash_and_paths(&self) -> (MerkleHash, Vec<MerklePath>) {
+        merklize(&self.parts.iter().map(|x| x.as_ref().unwrap()).collect::<Vec<_>>())
```
Comment (nearmax) [inline core/primitives/src/sharding.rs]: We should avoid `unwrap` on option in prod code.

File: core/primitives/src/test_utils.rs [hunk 13]
```diff
@@ -41,13 +55,13 @@ pub fn get_bls_key_pair_from_seed(seed_string: &str) -> (BlsPublicKey, BlsSecret
...
         InMemorySigner {
             account_id: account_id.to_string(),
             public_key,
             secret_key,
-            bls_public_key,
-            bls_secret_key,
+//            bls_public_key,
```
Comment (nearmax) [inline core/primitives/src/test_utils.rs]: Should we remove it?

File: core/primitives/src/traits.rs [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+/// ToBytes is like Into<Vec<u8>>, but doesn't consume self
+pub trait ToBytes: Sized {
+    fn to_bytes(&self) -> Vec<u8>;
+}
+
+pub trait Base64Encoded:
+    for<'a> TryFrom<&'a [u8], Error = String> + ToBytes
+{
+    fn from_base64(s: &str) -> Result<Self, String> {
+        let bytes = base64::decode(s).map_err(|e| format!("{}", e))?;
```
Comment (nearmax) [inline core/primitives/src/traits.rs]: `Box<std::error::Error>` instead of `String`.

File: core/primitives/src/utils.rs [hunk 15]
```diff
@@ -105,3 +106,30 @@ where
 {
     proto_to_result(proto).and_then(TryInto::try_into)
 }
+
+pub struct DisplayOption<T>(pub Option<T>);
```
Comment (nearmax) [inline core/primitives/src/utils.rs]: Add a comment explaining what it is.

File: core/store/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+                    write!(f, "  + {:?} {}\n", col, base64::encode(key))?
+                }
+                DBOp::Delete { col, key } => write!(f, "  - {:?} {}\n", col, base64::encode(key))?,
+            }
+        }
+        write!(f, "}}\n")
+    }
+}
+
+pub fn read_with_cache<'a, T: Decode + DeserializeOwned + std::fmt::Debug + 'a>(
```
Comment (nearmax) [inline core/store/src/lib.rs]: Why `T: Debug`?

File: test-utils/testlib/src/standard_test_cases.rs [hunk 17]
```diff
@@ -341,11 +342,11 @@ pub fn test_callback(node: RuntimeNode) {
 
     let mut state_update = node.client.read().unwrap().get_state_update();
     set(&mut state_update, key_for_callback(&callback_id), &callback);
-    let (root, transaction) = state_update.finalize();
+    let (transaction, root) = state_update.finalize();
```
Comment (nearmax) [inline test-utils/testlib/src/standard_test_cases.rs]: What is the reason to swap transaction and root?

**Comment by ilblackdragon** [comment]
@vans163 good suggestion, before switching would be good to understand support of it in other language libraries.
I'm going for now put call to base64 into single place to simplify switching to different encoding.

File: core/primitives/src/crypto/aggregate_signature.rs (hunk 9 above)
Comment (ilblackdragon) [reply core/primitives/src/crypto/aggregate_signature.rs]: What's the benefit of Box<std::error::Error>? I get custom error class with proper set of Froms to escalate errors, but why Box?

File: core/primitives/Cargo.toml (hunk 10 above)
Comment (ilblackdragon) [reply core/primitives/Cargo.toml]: Consistency.

File: core/primitives/src/crypto/signature.rs (hunk 11 above)
Comment (ilblackdragon) [reply core/primitives/src/crypto/signature.rs]:
```
    |
246 |         h.0.0.to_vec()
    |         --^^^
    |         | |
    |         | unexpected token
    |         help: try parenthesizing the first index: `(h.0).0`
```

File: core/primitives/src/crypto/aggregate_signature.rs (hunk 9 above)
Comment (nearmax) [reply core/primitives/src/crypto/aggregate_signature.rs]: Compatibility with third-party crates and std that almost always use structs that implement `std::error::Error`, e.g. io errors, network errors, etc. Boxing error allows us to pipe errors around the code without doing conversions. Additionally we can use `?` to make our code smaller. Without boxed error `?` will not work if some results inside the method are of different error type. Boxed error is a widespread pattern in Rust, enabled by `impl<T: Error> Error for Box<T>` in std https://doc.rust-lang.org/std/error/trait.Error.html . More info: https://doc.rust-lang.org/rust-by-example/error/multiple_error_types/boxing_errors.html

Comment (ilblackdragon) [reply core/primitives/src/crypto/aggregate_signature.rs]: Got it. Will do from now on.

**Comment by vans163** [comment]
@ilblackdragon I did a quick test it seems padding on/off is OK.  but uri safe base64 wont decode by regular base64 if it has a _ or a - in it.  I guess the point is ease of compatability, in which case uri safe would get confusing?

File: core/store/src/lib.rs (hunk 16 above)
Comment (ilblackdragon) [reply core/store/src/lib.rs]: good catch, was some debugging stuff.

File: test-utils/testlib/src/standard_test_cases.rs (hunk 17 above)
Comment (ilblackdragon) [reply test-utils/testlib/src/standard_test_cases.rs]: consistency, we were flipping them inside `finalize` before.

File: core/primitives/src/sharding.rs (hunk 12 above)
Comment (ilblackdragon) [reply core/primitives/src/sharding.rs]: Deleted this file - leaked from another pr.

**Comment by ilblackdragon** [comment]
Master:
```
test runtime_send_money                     ... bench:     130,352 ns/iter (+/- 22,359)
test runtime_wasm_bad_code                  ... bench:     143,678 ns/iter (+/- 64,006)
test runtime_wasm_benchmark_10_reads_legacy ... bench:     196,800 ns/iter (+/- 47,276)
test runtime_wasm_benchmark_storage_100     ... bench:     199,882 ns/iter (+/- 27,008)
test runtime_wasm_benchmark_storage_1000    ... bench:     201,110 ns/iter (+/- 36,142)
test runtime_wasm_benchmark_sum_1000        ... bench:     201,343 ns/iter (+/- 49,623)
test runtime_wasm_benchmark_sum_1000000     ... bench:     198,890 ns/iter (+/- 33,618)
test runtime_wasm_set_value                 ... bench:     208,993 ns/iter (+/- 146,110)
```

New:
```
running 8 tests
test runtime_send_money                     ... bench:     109,610 ns/iter (+/- 24,658)
test runtime_wasm_bad_code                  ... bench:     103,209 ns/iter (+/- 37,628)
test runtime_wasm_benchmark_10_reads_legacy ... bench:     221,626 ns/iter (+/- 80,122)
test runtime_wasm_benchmark_storage_100     ... bench:     221,655 ns/iter (+/- 23,496)
test runtime_wasm_benchmark_storage_1000    ... bench:     216,577 ns/iter (+/- 15,599)
test runtime_wasm_benchmark_sum_1000        ... bench:     217,897 ns/iter (+/- 13,979)
test runtime_wasm_benchmark_sum_1000000     ... bench:     219,221 ns/iter (+/- 16,854)
test runtime_wasm_set_value                 ... bench:     222,418 ns/iter (+/- 25,140)
```

File: core/primitives/src/logging.rs [hunk 18]
```diff
@@ -38,7 +37,7 @@ pub fn pretty_utf8(buf: &[u8]) -> String {
         Ok(s) => pretty_hash(s),
         Err(_) => {
             if buf.len() <= STRING_PRINT_LEN {
-                pretty_hash(&bs58::encode(buf).into_string())
+                pretty_hash(&base64::encode(buf))
```
Comment (bowenwang1996) [inline core/primitives/src/logging.rs]: Why not use `to_base64` here?

File: core/store/benches/trie_bench.rs [hunk 19]
```diff
@@ -44,7 +42,7 @@ fn trie_update(bench: &mut Bencher) {
 
     bench.iter(|| {
         let mut this_changes = changes.clone();
-        let (_, _) = trie.update(&root, this_changes.drain(..));
+        let _ = trie.update(&root, this_changes.drain(..));
```
Comment (bowenwang1996) [inline core/store/benches/trie_bench.rs]: Why is `let _` necessary?

File: core/store/src/lib.rs [hunk 20]
```diff
@@ -0,0 +1,188 @@
...
+                Ok(result) => Ok(Some(result)),
+                Err(e) => Err(e),
+            },
+            Ok(None) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn exists(&self, column: Option<u32>, key: &[u8]) -> Result<bool, io::Error> {
+        match self.storage.get(column, key) {
```
Comment (bowenwang1996) [inline core/store/src/lib.rs]: `self.storage.get(column, key).map(Option::is_some)`

File: core/store/src/lib.rs [hunk 21]
```diff
@@ -0,0 +1,188 @@
...
+
+/// Reads an object from Trie.
+pub fn get<T: DeserializeOwned>(state_update: &TrieUpdate, key: &[u8]) -> Option<T> {
+    state_update.get(key).and_then(|data| Decode::decode(&data).ok())
+}
+
+/// Writes an object into Trie.
+pub fn set<T: Serialize>(state_update: &mut TrieUpdate, key: Vec<u8>, value: &T) {
+    value.encode().ok().map(|data| state_update.set(key, DBValue::from_vec(data))).or_else(|| {
+        debug!("set value failed");
```
Comment (bowenwang1996) [inline core/store/src/lib.rs]: Do you still need this line?

File: core/store/src/trie/mod.rs [hunk 22]
```diff
@@ -393,17 +397,37 @@ pub struct TrieChanges {
...
-
 enum FlattenNodesCrumb {
     Entering,
     AtChild(Box<[Option<CryptoHash>; 16]>, usize),
     Exiting,
 }
 
+pub fn convert_to_store_update(trie: Arc<Trie>, changes: TrieChanges) -> (StoreUpdate, CryptoHash) {
+    let mut store_update = StoreUpdate::new_with_trie(trie.storage.store.storage.clone(), trie.clone());
+    let TrieChanges { old_root: _, new_root, insertions, deletions } = changes;
```
Comment (bowenwang1996) [inline core/store/src/trie/mod.rs]: Use `..` instead

File: core/store/benches/trie_bench.rs (hunk 19 above)
Comment (ilblackdragon) [reply core/store/benches/trie_bench.rs]: otherwise there is warning about unused value.

Comment (bowenwang1996) [reply core/store/benches/trie_bench.rs]: I see

File: core/primitives/src/utils.rs [hunk 23]
```diff
@@ -85,13 +82,42 @@ pub fn to_string_value(s: String) -> StringValue {
...
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.0 {
+            Some(ref v) => write!(f, "Some({})", v),
+            None => write!(f, "None"),
+        }
+    }
+}
+
+impl<T> DisplayOption<T> {
+    pub fn into(self) -> Option<T> {
```
Comment (nearmax) [inline core/primitives/src/utils.rs]: This should be a part of:
```
impl From<DisplayOption<T>> for Option<T> 
```

File: core/primitives/src/utils.rs [hunk 24]
```diff
@@ -85,13 +82,42 @@ pub fn to_string_value(s: String) -> StringValue {
...
+        self.0
+    }
+
+    pub fn as_ref(&self) -> &Option<T> {
```
Comment (nearmax) [inline core/primitives/src/utils.rs]: This should be a part of:
```
impl AsRef<Option<T>> for DisplayOption<T> 
```

File: core/store/src/lib.rs [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+                Ok(result) => Ok(Some(result)),
+                Err(e) => Err(e),
+            },
+            Ok(None) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    pub fn exists(&self, column: Option<u32>, key: &[u8]) -> Result<bool, io::Error> {
+        self.storage.get(column, key).map(|value| Option::is_some(&value))
```
Comment (nearmax) [inline core/store/src/lib.rs]:
```suggestion
        self.storage.get(column, key).map(|value| value.is_some())
```

File: core/store/src/lib.rs [hunk 26]
```diff
@@ -0,0 +1,180 @@
...
+    pub fn merge_transaction(&mut self, transaction: DBTransaction) {
+        for op in transaction.ops {
+            match op {
+                DBOp::Insert { col, key, value } => self.transaction.put(col, &key, &value),
+                DBOp::Delete { col, key } => self.transaction.delete(col, &key),
+            }
+        }
+    }
+
+    pub fn commit(self) -> Result<(), io::Error> {
```
Comment (nearmax) [inline core/store/src/lib.rs]: We need a command opposite to commit, e.g. abort that would purge trie cache and cancel transaction.

File: core/store/src/trie/mod.rs [hunk 27]
```diff
@@ -393,17 +393,37 @@ pub struct TrieChanges {
...
     Entering,
     AtChild(Box<[Option<CryptoHash>; 16]>, usize),
     Exiting,
 }
 
+pub fn convert_to_store_update(trie: Arc<Trie>, changes: TrieChanges) -> (StoreUpdate, CryptoHash) {
+    let mut store_update =
+        StoreUpdate::new_with_trie(trie.storage.store.storage.clone(), trie.clone());
+    let TrieChanges { new_root, insertions, deletions, .. } = changes;
+    for (key, value, rc) in insertions.into_iter() {
```
Comment (nearmax) [inline core/store/src/trie/mod.rs]:
```suggestion
    for (key, value, rc) in insertions {
```

File: core/store/src/lib.rs [hunk 28]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+pub fn create_store(path: &str) -> Arc<Store> {
+    let db_config = DatabaseConfig::with_columns(Some(NUM_COLS));
+    let db = Arc::new(Database::open(&db_config, path).expect("Failed to open the database"));
+    Arc::new(Store::new(db))
+}
+
+/// Reads an object from Trie.
+pub fn get<T: DeserializeOwned>(state_update: &TrieUpdate, key: &[u8]) -> Option<T> {
```
Comment (nearmax) [inline core/store/src/lib.rs]: Ideally getting stuff from state update should not require deserialization.

File: core/store/src/trie/mod.rs [hunk 29]
```diff
@@ -393,17 +393,37 @@ pub struct TrieChanges {
...
+        let storage_rc = trie.storage.retrieve_rc(&key).unwrap_or_default();
+        let bytes = RcTrieNode::encode(&value, storage_rc + rc).expect("Failed to serialize");
+        store_update.set(COL_STATE, key.as_ref(), &bytes);
+    }
+    for (key, value, rc) in deletions.into_iter() {
```
Comment (nearmax) [inline core/store/src/trie/mod.rs]:
```suggestion
    for (key, value, rc) in deletions {
```

File: core/primitives/src/utils.rs (hunk 23 above)
Comment (ilblackdragon) [reply core/primitives/src/utils.rs]: Outside of crate with Option.

File: core/store/src/lib.rs (hunk 26 above)
Comment (ilblackdragon) [reply core/store/src/lib.rs]: Will add it separately. Currently we would not have usage for it.

File: core/store/src/lib.rs (hunk 28 above)
Comment (ilblackdragon) [reply core/store/src/lib.rs]: ? just call `state_update.get(key)` then

File: core/primitives/src/utils.rs (hunk 23 above)
Comment (nearmax) [reply core/primitives/src/utils.rs]: I do not understand what do you mean.

Comment (nearmax) [reply core/primitives/src/utils.rs]: It does not have to be inside Option crate. It is normal to implement traits for types we don't own, as long as we own DisplayOption

**Comment by ilblackdragon** [comment]
Closing this as this heaps outdated.

**Pull request closed without merging**